File: Keynote/R/PageRank.R
Students = c("Alice", "Bob", "Chris", "Dylan", "Eve")
NumberOfFriends = c(1, 2, 2, 2, 1)
Love = c(1, 1, 1, 1, 1)
barplot(Love, ylim = c(0, 1.5), names.arg = Students)
alpha = 0.85

for (i in 1:80) {
    ShareTheLove = Love/NumberOfFriends
    # print(ShareTheLove)
    # left boundary, array index in R starts at 1.
    Love[1] = 1-alpha + alpha * ShareTheLove[2]
    # right boundary
    Love[5] = 1-alpha + alpha * ShareTheLove[4]
    # internal nodes
    Love[2:4] = 1-alpha + alpha * (ShareTheLove[1:3] + ShareTheLove[3:5])
    print(Love)
    barplot(Love, ylim = c(0, 1.5), names.arg = Students)
}

File: Keynote/Python/PageRank.py
import numpy as np

Student = ["Alice", "Bob", "Chris", "Dylan", "Eve"]
NumberOfFriends = np.array([1, 2, 2, 2, 1])
Love = np.array([1.0, 1.0, 1.0, 1.0, 1.0])

for i in range(1, 10):
  ShareTheLove = Love/NumberOfFriends
  # left boundary, array index starts from 0 in Python
  Love[0] = 0.15 + 0.85 * ShareTheLove[1]
  Love[4] = 0.15 + 0.85 * ShareTheLove[3]
  # internal nodes, Numpy array range does not include the "end",
  # i.e. array[0:3] => gives array elements 0, 1, 2
  Love[1:4] = 0.15 + 0.85 * (ShareTheLove[0:3] + ShareTheLove[2:5])
